File: javascript/anagram/anagram.js
var Anagram = function(word) {
  this.word = word;
};

Anagram.prototype.matches = function(matches) {
  if (typeof matches === 'string') { matches = [matches] }
  return matches.filter(this.isAnagram.bind(this));
};

Anagram.prototype.isAnagram = function(match) {
  var formattedWord = this.formatWord(this.word);
  var comparison = this.formatWord(match);
  return formattedWord === comparison && this.word.toLowerCase() !== match.toLowerCase();
};

Anagram.prototype.formatWord = function(word) {
  return word.toLowerCase().split('').sort().join(' ')
};

module.exports = Anagram;

File: java/anagram/src/main/java/Anagram.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class Anagram {

    private String input;

    public Anagram(String input) {
        this.input = input.toLowerCase();
    }

    public List<String> match(List<String> words) {
        List<String> anagrams = new ArrayList<>();
        for (String word : words) {
            if (isAnagram(word) && !isIdenticalToInput(word))  { anagrams.add(word); }
        }
        return anagrams;
    }

    private String sortLetters(String word) {
        char[] letters = word.toLowerCase().toCharArray();
        Arrays.sort(letters);
        return String.copyValueOf(letters);
    }

    private boolean isAnagram(String word) {
        return sortLetters(input).equals(sortLetters(word));
    }

    private boolean isIdenticalToInput(String word) {
        return input.equals(word.toLowerCase());
    }
}
